[NFC] Clarify import filtering logic and naming.

varungandhi-apple opened the pull request:
Ran into surprising behavior with `SourceFile::getImportedModules` when working on https://github.com/apple/swift/pull/33980, figured I should land this first.

I don't understand why `ModuleFile::getImportedModules` doesn't have handling for `SPI`/`ShadowedByCrossImportOverlay` though, so I haven't touched it.

varungandhi-apple commented:
@swift-ci test

varungandhi-apple commented:
https://github.com/apple/llvm-project/pull/1819

@swift-ci please test

varungandhi-apple commented:
@swift-ci please clean test Windows

varungandhi-apple commented:
apple/llvm-project#1819

@swift-ci please clean test Windows

varungandhi-apple commented:
The Windows failure is because Windows doesn't support cross-repository testing.

xymus reviewed (approved):
I was convinced at the first `s/Public/Exported/`. Thanks!

brentdax commented on lib/AST/Module.cpp:
```diff
@@ -1163,6 +1163,13 @@ void SourceFile::lookupPrecedenceGroupDirect(
 
 void ModuleDecl::getImportedModules(SmallVectorImpl<ImportedModule> &modules,
                                     ModuleDecl::ImportFilter filter) const {
+  assert((bool)(filter & ImportFilter({
```
Is there a method on OptionSet that expresses this more clearly than `&`? Just reads strangely.

brentdax commented on include/swift/AST/Module.h:
```diff
@@ -648,27 +648,52 @@ class ModuleDecl : public DeclContext, public TypeDecl {
-    ShadowedBySeparateOverlay = 1 << 4
+    /// Include imports shadowed by a cross-import overlay. Unshadowed imports
+    /// are included whether or not this flag is specified.
+    ShadowedByCrossImportOverlay = 1 << 4
```
Would something like `ReplacedByCrossImport` be even clearer?

varungandhi-apple replied:
I started implementing this and updating all the comments, but I feel like "shadowed" conveys the sense better? With shadowing, the old thing is still there, but the compiler is looking at something closer/more relevant. With "replace", it is less clear if/why the old thing sticks around.

brentdax replied:
Sure, that's fine.

beccadax reviewed (approved):
A definite improvement. (I think it’s this way because there was some thought that `@_exported import` might eventually be spelled `public import` when it became a supported language feature. But if that ever happens, we can cross that bridge when we get to it.)

hamishknight commented on include/swift/AST/Module.h:
```diff
@@ -648,27 +648,52 @@ class ModuleDecl : public DeclContext, public TypeDecl {
-    Public = 1 << 0,
+    Exported = 1 << 0,
     /// Include "regular" imports with no special annotation.
     Private = 1 << 1,
```
Would it be worth also renaming this case while we're here, to avoid confusion with `@_private` and `SourceFile::ImportFlags::PrivateImport`? Perhaps `Regular`?

varungandhi-apple replied:
Yeah, I was in two minds about this... I was going to name it `Regular` or `Default` but neither of them felt sufficiently descriptive. @brentdax , @xymus : any naming suggestions for this one?

xymus replied:
Even if it's not very descriptive I like `Default`, it's clear enough as being for the default import behavior.

brentdax replied:
`Default` is fine. `Interface` might work too (contrasting with `ImplementationOnly`).

hamishknight reviewed (commented):
Nice!

varungandhi-apple committed:
[NFC] Clarify semantics of getImportedModules.

The lack of clarity manifested as unexpected behavior when using
getImportedModules to create the module import graph. The new behavior
makes SPI-ness and Shadowing-ness behave similarly in terms of
filtering. We also check if a filter is well-formed to avoid
accidental empty import lists.

varungandhi-apple committed:
[NFC] Rename ImportFilterKind cases to be clearer.

varungandhi-apple commented:
apple/llvm-project#1819

@swift-ci please test
